miners/text/template_miner.py

```diff
@@ -0,0 +1,643 @@
+
+        # distillation_loss : distillation loss between local_context and remote_context
+        # distillation_loss.shape = [1]
+        # This trains the local_context (student) to emulate the network context.
+        output.distillation_loss = F.mse_loss( output.local_context, output.remote_hidden.detach() )
+
+        if training :
+            # remote_target: projection of remote_hidden onto target dimension.
+            # remote_target.shape = [batch_size, sequence_len, bittensor.__vocab_size__]
+            output.remote_target = self.remote_decoder( output.remote_hidden )
+
+            # remote_target_loss: MLM loss between remote_target and passed targets.
+            # remote_target_loss.shape = [1]
+            shift_logits = output.remote_target[..., :-1, :].contiguous()
+            shift_labels = inputs[..., 1:].contiguous()            
+            output.remote_target_loss = self.loss_fct( shift_logits.view(-1, shift_logits.size(-1)), shift_labels.view(-1) )
+        
+        return output
+
+    def remote(self, inputs: torch.int64 ) -> torch.float32:
+        """ Forwards the inputs through the network, selects the topk peers based on self.chain_weights.
+        Args:
+            inputs (:obj:`torch.int64` of shape :obj:`(batch_size, sequence_len)`, `required`): 
+                Batch_size length list of text sentences.
+        Returns:
+            outputs (:obj:`torch.FloatTensor` of shape :obj:`(batch_size, sequence_len, bittensor.__network_dim__)`, `optional`): 
+                Joined hidden layer responses from peers.
+        """
+        # ---- Topk Weights ---- (TODO: check if the gaussians are enough disrupt the chain weights)
+        real_topk = min( self.config.nucleus.topk, bittensor.neuron.metagraph.n.item() ) 
+        noise = torch.normal( torch.mean(self.chain_weights).item(), torch.std(self.chain_weights).item()+0.0000001, size=( self.chain_weights.size())).to( self.config.miner.device )
+        topk_weights, topk_uids = torch.topk( self.chain_weights + noise, real_topk, dim=0 ) 
+
+        # ---- Filter endpoints ----
+        endpoints = [bittensor.neuron.metagraph.endpoints[uid] for uid in topk_uids]
+
+        # ---- Query network ----
+        responses, return_ops = bittensor.neuron.dendrite.forward_text ( 
+            endpoints = endpoints, 
+            inputs = [inputs.long() for _ in endpoints]
+        )
+
+        # ---- Join based on weights ----
+        joining_weights = F.softmax( topk_weights, dim = 0 )
+        output = torch.zeros( (inputs.shape[0], inputs.shape[1], bittensor.__network_dim__)).to( self.config.miner.device )
+        for index, response in enumerate( responses ): 
+            output += response.to( self.config.miner.device ) * joining_weights[ topk_uids[index] ]
+
+        # ---- Punish peers with non-successful return ops ----
+        with torch.no_grad():
+            self.chain_weights[topk_uids[(return_ops != 0)]] -=  self.config.miner.punishment
```

Comment:
This arg should be part of the nucleus I think.

Reply:
Fixed!